spec/realtime/channel.test.js

```diff
@@ -336,6 +336,359 @@ define(['ably', 'shared_helper', 'async'], function(Ably, helper, async) {
+				var params = {
+					modes: 'subscribe',
+					delta: 'vcdiff'
+				};
+				var channel = realtime.channels.get(testName, {
+					params: params
+				});
+				channel.attach(function(err) {
+					if(err)
+						test.ok(false, 'Attach failed with error: ' + displayError(err));
+
+					try {
+						realtime.channels.get(testName, {
+							params: params
+						});
+					} catch(e) {
+						test.equal(e.code, 40000, 'Check error code');
+						test.equal(e.statusCode, 400, 'Check error status code');
+						test.ok(e.message.includes('setOptions'), 'Check error message');
+						closeAndFinish(test, realtime);
+					}
+				});
+			});
+			monitorConnection(test, realtime);
+		} catch(e) {
+			test.ok(false, testName + ' failed with exception: ' + e.stack);
+			closeAndFinish(test, realtime);
+		}
+	};
+
+	testOnAllTransports(_exports, 'setOptionsCallbackBehaviour', function(realtimeOpts) { return function(test) {
+		test.expect(6);
+		var testName = 'setOptionsCallbackBehaviour';
+		try {
+			var realtime = helper.AblyRealtime(realtimeOpts);
+			realtime.connection.on('connected', function() {
+				var params = {
+					modes: 'subscribe',
+					delta: 'vcdiff'
+				};
+				var modes = ['publish'];
+				var channel = realtime.channels.get(testName);
+
+				async.series([
+					function(cb) {
+						var setOptionsReturned = false;
+						channel.setOptions({
+							params: params,
+							modes: modes
+						}, function() {
+							test.ok(!setOptionsReturned, 'setOptions failed to call back immediately, when no reattach is required');
+							cb();
+						});
+						setOptionsReturned = true;
+					},
+					function(cb) {
+						channel.attach(function(err) {
+							cb(err);
+						});
+					},
+					function(cb) {
+						var channelReattached = false;
+						channel.on('attaching', function() {
+							channelReattached = true;
+						});
+
+						var setOptionsReturned = false;
+						channel.setOptions({
+							params: params
+						}, function() {
+							test.ok(setOptionsReturned, 'setOptions should return immediately and call back after the reattach');
+							test.ok(channelReattached, 'Check reattach with params occurred');
+							cb();
+						});
+						setOptionsReturned = true;
+					},
+					function(cb) {
+						var channelReattached = false;
+						channel.on('attaching', function() {
+							channelReattached = true;
+						});
+
+						var setOptionsReturned = false;
+						channel.setOptions({
+							modes: modes
+						}, function() {
+							test.ok(setOptionsReturned, 'setOptions should return immediately and call back after the reattach');
+							test.ok(channelReattached, 'Check reattach with modes occurred');
+							cb();
+						});
+						setOptionsReturned = true;
+					},
+					function(cb) {
+						var setOptionsReturned = false;
+						channel.setOptions({ }, function() {
+							test.ok(!setOptionsReturned, 'setOptions failed to call back immediately, when no reattach is required');
```

Comment:
As above, can we test that apart from the attributes being set on the client lib SDK channel, that this data is actually making it to Realtime.  I will stop adding this comment for subsequent tests :)

Reply:
Fixed in https://github.com/ably/ably-js/pull/613/commits/804461e067147114f10062235374bdf9b95a0c08